### PR TITLE
Hash table stats

### PR DESCRIPTION
**Pull request by jingjingwang**
Add REST APIs to Worker, including one that fetches hash table stats.

### CODE REVIEW

File: src/edu/washington/escience/myria/operator/SymmetricHashJoin.java [hunk 1]
```diff
@@ -515,4 +524,16 @@ public long getNumTuplesInHashTables() {
     /** Pull from the right child until it reaches EOS. */
     RIGHT_EOS
   }
+
+  @Override
+  public Map<String, Map<String, Integer>> dumpHashTableStats() {
+    Map<String, Map<String, Integer>> ret = new HashMap<String, Map<String, Integer>>();
```
Comment (senderista) [inline src/edu/washington/escience/myria/operator/SymmetricHashJoin.java]: Can you just use the diamond operator here?
```
 Map<String, Map<String, Integer>> ret = new HashMap<>();
```

File: src/edu/washington/escience/myria/operator/TupleHashTable.java [hunk 2]
```diff
@@ -129,4 +167,23 @@ public void cleanup() {
...
   }
+
+  /**
+   *
+   * @return stats of the hash table.
+   */
+  public Map<String, Integer> dumpStats() {
+    Map<String, Integer> ret = new HashMap<String, Integer>(stats);
+    ret.put("numTuples", numTuples());
+    ret.put("numKeys", keyHashCodesToIndices.size());
```
Comment (senderista) [inline src/edu/washington/escience/myria/operator/TupleHashTable.java]: Of course this isn't quite accurate once the number of keys approaches 2^16...

File: src/edu/washington/escience/myria/operator/agg/Aggregate.java [hunk 3]
```diff
@@ -219,5 +221,15 @@ protected void init(final ImmutableMap<String, Object> execEnvVars) throws DbExc
         new UniqueTupleHashTable(
             Schema.merge(groupingSchema, stateSchema), MyriaArrayUtils.range(0, gfields.length));
     resultBuffer = new TupleBatchBuffer(getSchema());
+    groupStates.name = "op" + getOpId();
+  }
+
+  @Override
+  public Map<String, Map<String, Integer>> dumpHashTableStats() {
+    Map<String, Map<String, Integer>> ret = new HashMap<String, Map<String, Integer>>();
```
Comment (senderista) [inline src/edu/washington/escience/myria/operator/agg/Aggregate.java]: again I think you can use the diamond operator

File: src/edu/washington/escience/myria/parallel/LocalFragment.java [hunk 4]
```diff
@@ -931,4 +937,18 @@ public int getMaxOpId(final Operator op) {
...
     return ret;
   }
+
+  /**
+   *
+   * @param op the operator
+   * @return the hash table stats of the operators that are children of op
+   */
+  public Map<String, Map<String, Integer>> dumpHashTableStats(final Operator op) {
+    Map<String, Map<String, Integer>> ret = new HashMap<String, Map<String, Integer>>();
```
Comment (senderista) [inline src/edu/washington/escience/myria/parallel/LocalFragment.java]: again I think you can use the diamond operator

File: src/edu/washington/escience/myria/parallel/WorkerSubQuery.java [hunk 5]
```diff
@@ -351,4 +353,15 @@ private int getMaxOpId() {
     }
     return ret;
   }
+
+  /**
+   * @return the hash table stats of the subquery.
+   */
+  public Map<String, Map<String, Integer>> dumpHashTableStats() {
+    Map<String, Map<String, Integer>> ret = new HashMap<String, Map<String, Integer>>();
```
Comment (senderista) [inline src/edu/washington/escience/myria/parallel/WorkerSubQuery.java]: again I think you can use the diamond operator

File: src/edu/washington/escience/myria/parallel/Worker.java [hunk 6]
```diff
@@ -840,4 +862,15 @@ public synchronized PythonFunctionRegistrar getPythonFunctionRegistrar() throws
     }
     return pythonFunctionRegistrar;
   }
+
+  /**
+   * @return the working directory of the Myria worker.
+   */
+  /*
```
Comment (senderista) [inline src/edu/washington/escience/myria/parallel/Worker.java]: why is this commented out?

File: src/edu/washington/escience/myria/tools/MyriaConfigurationParser.java [hunk 7]
```diff
@@ -323,6 +325,46 @@ private static int getPort(final ConfigParser parser, final int nodeId)
 
   /**
    *
+   * @param nodeId the worker/master ID
+   * @return the port number
+   * @throws ConfigFileException if error occurred when getting the value
+   */
+  private static int getJVMPort(final ConfigParser parser, final int nodeId)
```
Comment (senderista) [inline src/edu/washington/escience/myria/tools/MyriaConfigurationParser.java]: Doesn't the JVM port only exist in your hacked OpenJDK version? What happens if a user specifies a nonzero value for this port on a standard JVM?

File: src/edu/washington/escience/myria/tools/MyriaWorkerConfigurationModule.java [hunk 8]
```diff
@@ -25,6 +25,8 @@
   public static final RequiredParameter<Integer> WORKER_ID = new RequiredParameter<>();
   public static final RequiredParameter<String> WORKER_HOST = new RequiredParameter<>();
   public static final RequiredParameter<Integer> WORKER_PORT = new RequiredParameter<>();
+  public static final OptionalParameter<Integer> WORKER_JVM_PORT = new OptionalParameter<>();
```
Comment (senderista) [inline src/edu/washington/escience/myria/tools/MyriaWorkerConfigurationModule.java]: see above

File: src/edu/washington/escience/myria/tools/MyriaWorkerConfigurationModule.java [hunk 9]
```diff
@@ -52,6 +56,12 @@
   public class WorkerPort implements Name<Integer> {}
 
   @NamedParameter
+  public class WorkerStatsPort implements Name<Integer> {}
+
+  @NamedParameter
+  public class WorkerJVMPort implements Name<Integer> {}
```
Comment (senderista) [inline src/edu/washington/escience/myria/tools/MyriaWorkerConfigurationModule.java]: see above

File: src/edu/washington/escience/myria/expression/LikeExpression.java [hunk 10]
```diff
@@ -0,0 +1,28 @@
+package edu.washington.escience.myria.expression;
+
+import edu.washington.escience.myria.SimplePredicate;
+
+/**
+ * Like Expression.
+ */
+public class LikeExpression extends ComparisonExpression {
```
Comment (senderista) [inline src/edu/washington/escience/myria/expression/LikeExpression.java]: see above, I don't understand the motivation/application of this expression.

File: src/edu/washington/escience/myria/daemon/MyriaDriver.java [hunk 11]
```diff
@@ -590,6 +591,17 @@ private void setJVMOptions(final AllocatedEvaluator evaluator, final boolean isM
                         MyriaGlobalConfigurationModule.WorkerJvmHeapSizeMaxGB.class)));
     final Set<String> jvmOptions =
         globalConfInjector.getNamedInstance(MyriaGlobalConfigurationModule.JvmOptions.class);
+    if (workerConf != null && jvmOptions.contains("-ElasticMem")) {
```
Comment (senderista) [inline src/edu/washington/escience/myria/daemon/MyriaDriver.java]: What if someone specifies `-ElasticMem` on a standard JVM?

File: src/edu/washington/escience/myria/expression/BinaryExpression.java [hunk 12]
```diff
@@ -89,15 +89,25 @@ protected final String getInfixBinaryString(
...
-    return new StringBuilder("(")
-        .append(getLeft().getJavaString(parameters))
-        .append(".compareTo(")
-        .append(getRight().getJavaString(parameters))
-        .append(')')
-        .append(op.toJavaString())
-        .append(0)
-        .append(")")
-        .toString();
+    if (op == SimplePredicate.Op.LIKE) {
```
Comment (senderista) [inline src/edu/washington/escience/myria/expression/BinaryExpression.java]: Also, regardless of the merits of this `LIKE` implementation, what is it doing in this PR?

File: src/edu/washington/escience/myria/api/WorkerResource.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+public final class WorkerResource {
+  /** The Myria server running on the worker. */
+  @Context private Worker worker;
+
+  /**
+   * @param qid query id.
+   * @return the stats of all hash tables of the given query.
+   */
+  @GET
+  @Path("/hashtable-query-{qid:\\d+}")
```
Comment (senderista) [inline src/edu/washington/escience/myria/api/WorkerResource.java]: I think a path format like `/query/<QID>/hashtable-stats` might be more readable.

File: src/edu/washington/escience/myria/daemon/MyriaDriver.java (hunk 11 above)
Comment (jingjingwang) [reply src/edu/washington/escience/myria/daemon/MyriaDriver.java]: Then the JVM would crash with something like: this JVM arg is unknown.

File: src/edu/washington/escience/myria/tools/MyriaConfigurationParser.java (hunk 7 above)
Comment (jingjingwang) [reply src/edu/washington/escience/myria/tools/MyriaConfigurationParser.java]: See above.

File: src/edu/washington/escience/myria/expression/BinaryExpression.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (jingjingwang) [reply src/edu/washington/escience/myria/expression/BinaryExpression.java]: `LIKE` is for TPC-H. I've modified it to be the SQL "like".

**Pull request merged**